### PR TITLE
Issue #279: Fix FSharpFunc unit type args

### LINKED ISSUE #279: Problems with lambdas of the form (_ -> unit)
### Description

Generated code involving lambdas of the form `_ -> unit` causes issues.

### Repro steps

Add (using the type provider template)
```fsharp 
        let meth = ProvidedMethod("StaticMethod", [], typeof<int []>, isStatic=true, invokeCode = (fun args -> <@@  let a = [|0 .. 10|] in a |> Array.iteri (fun i x -> a.[i] <- x + 1); a @@> ))
        myType.AddMember(meth)
```
to `BasicGenerativeProvider`.

Execute in fsi 
```fsharp
type Bleh = LemonadeProvider.GenerativeProvider<3>
Bleh.StaticMethod()
```

### Expected behavior

Return the array `[|1 .. 11|]`.

### Actual behavior

Exception:

```fsharp
> System.TypeLoadException: The signature is incorrect.
   at LemonadeProvider.Bleh.StaticMethod()
   at <StartupCode$FSI_0006>.$FSI_0006.main@() in E:\Script6.fsx:line 6
Stopped due to error
```

### Known workarounds

Avoid `'a -> unit`

### Related information 

Comparing the generated assembly with valid code it seems the issue is the generated lambda class inherits from `FSharpFunc<_,void>` as  oppose to the valid code inheriting from `FSharpFunc<_,unit>`. For example 
```csharp
using Microsoft.FSharp.Core;
using System.Runtime.InteropServices;

internal class Lambdadd375771\u002D88ae\u002D46e9\u002D92b2\u002D100258dec3aa : FSharpFunc<int, void>
{
  public Lambdadd375771\u002D88ae\u002D46e9\u002D92b2\u002D100258dec3aa()
  {
    base.\u002Ector();
  }

  public override sealed void Invoke([In] int obj0)
  {
    Operators.Ignore<int>((int) (obj0 + 21));
  }
}
```

### PR DESCRIPTION
**Pull request by kevmal**
Fixes #279

Problem is `emitLambda` relies on `body.Type` to build the `FSharpFunc` type. `body.Type` in the case of a `_ -> unit` function is `System.Void` so the type getting built is `FSharpFunc<_,void>`. Meanwhile a function of the form `unit -> _` is fine since the first type argument uses the type of the lambda var.

I resorted to using `FullName` on the generic type definitions to check for `FSharpFunc` type, maybe there's a better way?

### CODE REVIEW

File: src/ProvidedTypes.fs [hunk 1]
```diff
@@ -13729,7 +13729,21 @@ namespace ProviderImplementation.ProvidedTypes
...
             let lambda: ILTypeBuilder = assemblyMainModule.DefineType(UNone, genUniqueTypeName(), TypeAttributes.Class)
-            let baseType = convTypeToTgt (typedefof<FSharpFunc<_, _>>.MakeGenericType(v.Type, body.Type))
+
+            let fsharpFuncType = convTypeToTgt (typedefof<FSharpFunc<_, _>>)
+            let voidType = convTypeToTgt typeof<System.Void>
+            let rec lambdaType (t : Type) = 
+                if t.IsGenericType then 
+                    let args = t.GetGenericArguments()
+                    let gdef = t.GetGenericTypeDefinition()
+                    if args.Length = 2 && gdef.FullName = fsharpFuncType.FullName && args.[1] = voidType then
```
Comment (baronfel) [inline src/ProvidedTypes.fs]: you can use [`FSharp.Reflection.FSharpType.IsFunction`](https://msdn.microsoft.com/visualfsharpdocs/conceptual/fsharptype.isfunction-method-%5bfsharp%5d) to test if a type is a F# function, and [`FSharp.Reflection.FSharpType.GetFunctionElements`](https://msdn.microsoft.com/visualfsharpdocs/conceptual/fsharptype.getfunctionelements-method-%5bfsharp%5d) to get the types of the domain and range of the function, checking those for the approproate matching types.

Comment (kevmal) [reply src/ProvidedTypes.fs]: The type has already been converted to the target context

Comment (kevmal) [reply src/ProvidedTypes.fs]: Admittedly I just assumed it wouldn't work and didn't try it. But I ran a quick test to confirm and I get `Message: System.NotSupportedException `

**Comment by dsyme** [comment]
Could you add a corresponding test case for this please? THanks

**Comment by kevmal** [comment]
Added some tests. Also there's a skipped test for #323.

**Pull request merged**